components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt

```diff
@@ -0,0 +1,338 @@
+                else -> Log.e(LOG_TAG, "Invalid $action (state: $currentState)")
+            }
+        }
+    }
+
+    internal suspend fun sendEvent(event: FxaEvent) {
+        Log.d(LOG_TAG, "Sending $event")
+        try {
+            eventHandler.onFxaEvent(event)
+        } catch (e: Exception) {
+            Log.e(LOG_TAG, "Exception sending $event", e)
+        }
+    }
+
+    private suspend fun changeState(newState: FxaAuthState, transition: FxaAuthStateTransition) {
+        Log.d(LOG_TAG, "Changing state from ${currentState} to ${newState}")
+        currentState = newState
+        sendEvent(FxaEvent.AuthStateChanged(newState, transition))
+    }
+
+    // Perform an operation, retrying after network errors
+    private suspend fun<T> withNetworkRetry(operation: suspend () -> T): T {
+        while (true) {
+            try {
+                if (simulateNetworkErrorFlag) {
+                    simulateNetworkErrorFlag = false
+                    throw FxaException.Network("Simulated Error")
+                }
+                return operation()
+            } catch(e: FxaException.Network) {
+                if (retryLogic.shouldRetryAfterNetworkError()) {
+                    Log.d(LOG_TAG, "Network error: retrying")
+                    continue
+                } else {
+                    Log.d(LOG_TAG, "Network error: not retrying")
+                    throw e
+                }
+            }
+        }
+    }
+
+    // Perform an operation, retrying after network errors and calling checkAuthorizationStatus
+    // after auth errors
+    private suspend fun<T> withRetry(operation: suspend () -> T): T {
+        while (true) {
+            try {
+                return withNetworkRetry(operation)
+            } catch (e: FxaException.Authentication) {
+                val currentState = currentState
+
+                if (currentState !is FxaAuthState.Connected) {
+                    throw e
+                }
+
+                if (retryLogic.shouldRecheckAuthStatus()) {
+                    Log.d(LOG_TAG, "Auth error: re-checking")
+                    handleCheckAuthorization(currentState)
+                } else {
+                    Log.d(LOG_TAG, "Auth error: disconnecting")
+                    inner.disconnectFromAuthIssues()
+                    tryPersistState()
+                    changeState(FxaAuthState.Disconnected(true), FxaAuthStateTransition.AUTH_CHECK_FAILED)
+                }
+
+                if (this.currentState is FxaAuthState.Connected) {
+                    continue
+                } else {
+                    throw e
+                }
+            }
+        }
+    }
+
+    private suspend fun handleBeginOAuthFlow(currentState: FxaAuthState.Disconnected, action: FxaAction.BeginOAuthFlow) {
+        handleBeginEitherOAuthFlow(currentState, action.result) {
+            inner.beginOauthFlow(action.scopes.toList(), action.entrypoint, MetricsParams(mapOf()))
```

Comment:
I'm I handling `MetricsParams` correctly?  I don't really understand how it's supposed to work.

Reply:
It gets passed in from the caller - it includes query parameters that the client wants to append to the URL, eg: all the `utm_*` params in https://mozilla.github.io/ecosystem-platform/relying-parties/reference/query-parameters#generic-parameters

Reply:
Ahh okay.  I don't think Android uses this, but I added this as an optional parameter for the begin oauth actions.